gateway/loaders/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,343 @@
+		Request:    requestMap,
+	}
+
+	return t, nil
+}
+
+func (z *ZookeeperTargetLoader) zookeeperToTargets(t *TargetConfig) (*targetpb.Configuration, error) {
+
+	configs := &targetpb.Configuration{
+		Target:  make(map[string]*targetpb.Target),
+		Request: make(map[string]*gnmi.SubscribeRequest),
+	}
+
+	for requestName, request := range t.Request {
+		var subs []*gnmi.Subscription
+		for _, x := range request.Paths {
+
+			var origin = ""
+			split := strings.Split(x, ":")
+			if len(split) > 1 {
+				origin = split[0]
+				x = split[1]
+			}
+
+			path, err := xpath.ToGNMIPath(x)
+			if err != nil {
+				return nil, fmt.Errorf("unable to parse zookeeper config XPath: %v", err)
+			}
+			if origin != "" {
+				path.Origin = origin
+			}
+			subs = append(subs, &gnmi.Subscription{Path: path})
+		}
+		configs.Request[requestName] = &gnmi.SubscribeRequest{
+			Request: &gnmi.SubscribeRequest_Subscribe{
+				Subscribe: &gnmi.SubscriptionList{
+					Prefix: &gnmi.Path{
+						Target: request.Target,
+					},
+					Subscription: subs,
+				},
+			},
+		}
+	}
+
+	for connName, conn := range t.Connection {
+		configs.Target[connName] = &targetpb.Target{
+			Addresses: conn.Addresses,
+			Request:   conn.Request,
+			Meta:      conn.Meta,
+			Credentials: &targetpb.Credentials{
+				Username: conn.Credentials.Username,
+				Password: conn.Credentials.Password,
+			},
+		}
+	}
+	return configs, nil
+}
+
+func (z *ZookeeperTargetLoader) GetConfiguration() (*targetpb.Configuration, error) {
+	zClient, err := NewZookeeperClient(z.hosts, z.config.ZookeeperTimeout)
+	if err != nil {
+		return nil, err
+	}
+	data, err := zClient.GetZookeeperData()
+	switch err {
+	case zk.ErrNoNode:
+		configs := &targetpb.Configuration{}
+		z.config.Log.Info().Msg("Could not parse zookeeper configuration: " + err.Error())
+		if exists, _, _ := zClient.conn.Exists(CleanPath(TargetPath)); !exists {
+			z.config.Log.Info().Msg("Creating path: " + TargetPath)
+			if err := CreateFullPath(zClient.conn, TargetPath, zk.WorldACL(zk.PermAll)); err != nil {
+				return nil, err
+			}
+		}
+		if exists, _, _ := zClient.conn.Exists(CleanPath(RequestPath)); !exists {
+			z.config.Log.Info().Msg("Creating path: " + RequestPath)
+			if err := CreateFullPath(zClient.conn, RequestPath, zk.WorldACL(zk.PermAll)); err != nil {
+				return nil, err
+			}
+		}
+		return configs, nil
+	case nil:
+		z.config.Log.Info().Msg("Succesfully retreived zookeeper configuration")
+	default:
+		return nil, err
+	}
+
+	configs, err := z.zookeeperToTargets(data)
+
+	if err := target.Validate(configs); err != nil {
+		return nil, fmt.Errorf("configuration in zookeeper is invalid: %v", err)
+	}
+
+	return configs, nil
+}
+
+func (z *ZookeeperTargetLoader) Start() error {
+	_, err := z.GetConfiguration() // make sure there are no errors at startup
+	return err
+}
+
+func (z *ZookeeperTargetLoader) WatchConfiguration(targetChan chan<- *connections.TargetConnectionControl) error {
+	for {
```

Comment:
We're polling Zookeeper, which is inefficient. Can we use Zookeeper notifications instead? https://zookeeper.apache.org/doc/r3.3.3/zookeeperProgrammers.html#ch_zkWatches